src/Compilers/CSharp/Portable/Syntax/AnonymousFunctionExpressionSyntax.cs

```diff
@@ -23,5 +23,33 @@ public AnonymousFunctionExpressionSyntax WithAsyncKeyword(SyntaxToken asyncKeywo
             => WithAsyncKeywordCore(asyncKeyword);
 
         internal abstract AnonymousFunctionExpressionSyntax WithAsyncKeywordCore(SyntaxToken asyncKeyword);
+
+        /// <summary>
+        /// If the given <paramref name="asyncKeyword"/> is default, remove async if exists.
+        /// Otherwise, replace the existing <see cref="AsyncKeyword"/> or add a new one.
+        /// </summary>
+        private protected SyntaxTokenList UpdateAsyncKeyword(SyntaxToken asyncKeyword)
+        {
+            var existingAsync = AsyncKeyword;
```

Comment:
>AsyncKeyword [](http://example.com/codeflow?start=32&length=12)

Should we be operating on ```Modifiers``` instead? For example, if we are asked to remove an ```async``` keyword, then, perhaps, we should remove all of them from ```Modifiers```, not just the first one.

Reply:
@AlekseyTs My expectation of the `WithAsyncKeyword` is to be consistent with the behavior of `AsyncKeyword` property.

Since it's just "AsyncKeyword" that returns a single token, rather than "AsyncKeywords", I'm expecting `WithAsyncKeyword` to replace whatever token "AsyncKeyword" returns.

Reply:
> My expectation of the WithAsyncKeyword is to be consistent with the behavior of AsyncKeyword property.

> Since it's just "AsyncKeyword" that returns a single token, rather than "AsyncKeywords", I'm expecting WithAsyncKeyword to replace whatever token "AsyncKeyword" returns.

Do we have tests demonstraiting behavior of AsyncKeyword property after the replacement in presence of multiple keywords in modifiers? Does it really return ```default``` after ```UpdateAsyncKeyword(default)```? Does it really return the token that was passed in in other cases?

Reply:
> in presence of multiple keywords in modifiers? Does it really return `default` after `UpdateAsyncKeyword(default)`?

It will return the second async token in the original node (which will be the first of the new node). I see the point, and I'm a bit confused what the behavior should be. One possible approach is to operate only on the first async token when we are "adding" the token or "replacing" an existing one, but operate on all when removing. With this approach, accessing `.AsyncKeyword` after `UpdateAsyncKeyword` should always result in whatever was passed in.

> Does it really return the token that was passed in in other cases?

Yes I believe this should be the case, will update the tests after you confirm what the expected behavior of the API should be

Reply:
> will update the tests after you confirm what the expected behavior of the API should be

I believe it was a mistake in the first place to add a With-er for a calculated property, i.e. property that is effectively readonly and isn't part of the node description in syntax.xml. But since we are already in this uncomfortable position, I believe the API should always remove all async keywords and, if a replacement is supplied, it should be placed at the position where the first async keyword used to be. But this is just my opinion, others might have different ideas.

Reply:
> I believe it was a mistake in the first place to add a With-er for a calculated property

Yeah.  This was an aspect of how the API morphed over time.  We initially only supported (And thus exposed) the 'AsyncKeyword' of these nodes.  Then, we moved to a token-list because we wanted to support `static async` or `async static`.  In the migration, for binary compat, we had to preserve the AsyncKeyword property (and we didn't want it to throw or return 'default'), so we had it fetch out the keyword from teh token list if present.  

Sounds like i rushed this part of that syntax work, and this type of issue was the fallout.  Next time i'll pay more close attention to when we switch forms and that causes this.

That said, here's how i think it should work:

1. AsyncKeyword being *read* should still fetch out the keyword from teh token list if present.
2. WithAsyncKeyword should be deprecated and we should tell people to use the token list.

What WithAsyncKeyword should do when there already is such a keyword is up in the air.  However, i lean toward being somewhat sensible.  So if there is an existing async keyword, it gets removed (following the logic of all other withers).  *Where* the new AsyncKeyword shoudl go is more flexible.  I'm ok with it always going at the start (or end) of the token list.  I think such uniform logic is sensible and fits with the 'syntax doesn't try to have heuristics in it'.

Reply:
> WithAsyncKeyword should be deprecated and we should tell people to use the token list.

For now I only fixed implementation. Should I open an issue for discussion about deprecating this API?